### PR TITLE
Update E0102's example (fixes #33057)

### LINKED ISSUE #33057: E0102 explanantion has out of date / incorrect example.
`rustc --explain E0102` 

Gives this example code to show the problem.  
```rust
fn demo(devil: fn () -> !) {
    let x: &_ = devil();
    // error: cannot determine a type for this local variable
}

fn oh_no() -> ! { panic!("the devil is in the details") }

fn main() {
    demo(oh_no);
}
```
This example does not return E0102, but E0282. 

```
rustc --version --verbose
rustc 1.6.0 (c30b771ad 2016-01-19)
binary: rustc
commit-hash: c30b771ad9d44ab84f8c88b80c25fcfde2433126
commit-date: 2016-01-19
host: x86_64-unknown-linux-gnu
release: 1.6.0
```

### PR DESCRIPTION
**Pull request by Manishearth**
r? @GuillaumeGomez

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -1420,45 +1420,24 @@ fn main() {
...
+    // could be an array of anything
+    let x = []; // error: cannot determine a type for this local variable
 }
 ```
 
 To solve this situation, constrain the type of the variable.
 Examples:
 
-```no_run
+```rust
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: rust isn't needed in this case as far as I know. If no option is passed, the code is compiled and run.

**Comment by GuillaumeGomez** [comment]
Except for the nit (not really one but I like to keep consistent if the other error codes examples), it's all good. Thanks!

**Comment by GuillaumeGomez** [comment]
Fixes #33057.

File: src/librustc_typeck/diagnostics.rs (hunk 1 above)
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: Yeah I like to be explicit about this.

Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: If we keep it, it'll be the only place where such a flag is used. Also, by default, it is rust code. We only precise it when it's not.

Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: Done

**Comment by Manishearth** [comment]
@bors r=GuillaumeGomez rollup

@GuillaumeGomez you need to have `r+` somewhere in the command

**Pull request merged**